stereo_image_proc/src/stereo_image_proc/disparity_node.cpp

```diff
@@ -258,7 +259,10 @@ DisparityNode::DisparityNode(const rclcpp::NodeOptions & options)
   this->declare_parameters("", double_params);
   this->declare_parameters("", bool_params);
 
-  pub_disparity_ = create_publisher<stereo_msgs::msg::DisparityImage>("disparity", 1);
+  const auto disparity_pub_qos = use_system_default_qos ?
+    rclcpp::SystemDefaultsQoS() : rclcpp::QoS(1);
```

Comment:
Why ` rclcpp::QoS(1)` and not `rclcpp::SensorDataQoS()`? I see the existing code was just the `1`, but if we're exposing this option, shouldn't it be `SensorDataQoS`?

Reply:
Ci is working now too, oddly

Reply:
`rclcpp::QoS(1)` is equivalent to just passing `1`, which is using the default QoS with depth = 1. This actually has different settings than the `SensorDataQoS()`.

Reply:
I understand that, I'm asking for that change to be made to `SensorDataQoS` to complement the subscribers for consistency (that we should have had before)

Reply:
Ah, sorry, misunderstood.

In the original port, I consciously left the disparity pub/sub pair (and the point cloud publisher) reliable by default as it was closest to the ROS 1 behaviour. But, seeing we've already changed the image subscribers to best effort, I don't mind changing all pubs and subs to use best effort by default if that's your preference.

Reply:
Well, actually, I think setting the publishers reliability to *reliable* is the most flexible, since it is compatible with both *reliable* and *best effort* subscriptions.

Reply:
I'm thinking about in a pipeline, if you have this as step 7 / 10, you'd want to have the same input and outputs for each (e.g. if you wanted to subscribe to unreliable SensorDataQoS, you'd expect the param to also mess with the publisher so that they're all in line). 

With that said, I understand your comment and I think that's a valid point as well. I'm not really sure then what to do here. The 3 options we've mentioned:
- Publisher just being QoS(1) [before PR]
- Publisher having a param for the same as Subscriber for SensorData vs SystemDefaults [my original suggestion]
- Publisher having a param but with QoS(1) as the alternative [what's there now].

I would think for sensor processing the assumption would be defaulting to a sensor data QoS, but I suppose the question would be: in a pipeline of processing data, at what point should we switch from unreliable to reliable (or should we at all) between stages? Only first stage is unreliable and rest are reliable seems to be a reasonable choice, but so does all being unreliable so you don't lag behind if one stage has an issue doesn't try to retransmit.

Reply:
I'll say that I think any decision is better than what's there now. I don't think we need to go too overboard in deciding this. There should be an image-pipeline-full analysis of QoS and making it all consistent later.

Reply:
>  I suppose the question would be: in a pipeline of processing data, at what point should we switch from unreliable to reliable (or should we at all) between stages?

I think this is the question. And I'd posit the answer is different per application; we're not going to find an ideal configuration for everyone. By adding an option to fall back to the RMW configuration (system default), we give the user full control over how they'd like to configure the pipeline.

For default behaviour, I think making the nodes in the pipeline the most flexible (most compatible with other QoS settings) is a reasonable choice. For reliability, this means defaulting all subscriptions to best effort and all publishers to reliable. I think this is the current state of the PR.

Reply:
I'm not sure I agree since the default config or Qos(1) are incredibly similar. I think that's a misuse of the variable for `use_system_default_qos` since the opposite for a subscriber is a sensordataQoS and the publisher is essentially the same thing.

I'd merge this if we just removed the use of `use_system_default_qos` for the publisher entirely and just defaulted to system default for that (for now).

Reply:
Maybe we have a different view of `use_system_default_qos`. To me it means the user is responsible for configuring the QoS directly with the RMW. I don't think it's safe to assume the ROS default settings (`QoS(1)`) will be similar to the RMW default.

I've removed the use of the flag for the publisher for now (9827cbe).

If we change the publishers to default to SensorDataQoS in the future, I suggest releasing the change into a new ROS distro because it will break out-of-the-box experience for users. E.g. point clouds in RViz will not render without changes to the RViz point cloud display QoS configuration.

Reply:
Yeah, fair enough. We're still in porting mode at the moment. We're nearing the point where we can do a release but its currently unreleased in any ROS2 distribution as of date.